blockchain/storage/leveldb/builder.go

```diff
@@ -1,156 +1,30 @@
-	builder.options.WriteBuffer = (cacheSize / 4) * leveldbopt.MiB
-
-	builder.logger.Info("leveldb",
-		"BlockCacheCapacity", fmt.Sprintf("%d Mib", cacheSize/2),
-		"WriteBuffer", fmt.Sprintf("%d Mib", cacheSize/4),
-	)
-
-	return builder
+type leveldbStorageBuilder struct {
+	logger         hclog.Logger
+	leveldbBuilder kvdb.LevelDBBuilder
 }
 
-func (builder *leveldbBuilder) SetHandles(handles int) Builder {
-	builder.options.OpenFilesCacheCapacity = max(handles, minHandles)
-
-	builder.logger.Info("leveldb",
-		"OpenFilesCacheCapacity", builder.options.OpenFilesCacheCapacity,
-	)
-
-	return builder
-}
-
-func (builder *leveldbBuilder) SetBloomKeyBits(bloomKeyBits int) Builder {
-	builder.options.Filter = filter.NewBloomFilter(bloomKeyBits)
-
-	builder.logger.Info("leveldb",
-		"BloomFilter bits", bloomKeyBits,
-	)
-
-	return builder
-}
-
-func (builder *leveldbBuilder) SetCompactionTableSize(compactionTableSize int) Builder {
-	builder.options.CompactionTableSize = compactionTableSize * leveldbopt.MiB
-
-	builder.logger.Info("leveldb",
-		"CompactionTableSize", fmt.Sprintf("%d Mib", compactionTableSize),
-	)
-
-	return builder
-}
-
-func (builder *leveldbBuilder) SetCompactionTotalSize(compactionTotalSize int) Builder {
-	builder.options.CompactionTotalSize = compactionTotalSize * leveldbopt.MiB
-
-	builder.logger.Info("leveldb",
-		"CompactionTotalSize", fmt.Sprintf("%d Mib", compactionTotalSize),
-	)
-
-	return builder
-}
-
-func (builder *leveldbBuilder) SetNoSync(noSync bool) Builder {
-	builder.options.NoSync = noSync
-
-	builder.logger.Info("leveldb",
-		"NoSync", noSync,
-	)
-
-	return builder
-}
-
-func (builder *leveldbBuilder) Build() (storage.Storage, error) {
-	db, err := leveldb.OpenFile(builder.path, builder.options)
+func (builder *leveldbStorageBuilder) Build() (storage.Storage, error) {
+	db, err := builder.leveldbBuilder.Build()
 	if err != nil {
 		return nil, err
 	}
 
-	kv := &levelDBKV{db}
-
-	return storage.NewKeyValueStorage(builder.logger.Named("leveldb"), kv), nil
+	return storage.NewKeyValueStorage(builder.logger.Named("leveldb"), db), nil
 }
 
-// NewBuilder creates the new leveldb storage builder
-func NewBuilder(logger hclog.Logger, path string) Builder {
-	return &leveldbBuilder{
-		logger: logger,
-		path:   path,
-		options: &leveldbopt.Options{
-			OpenFilesCacheCapacity: minHandles,
-			CompactionTableSize:    DefaultCompactionTableSize * leveldbopt.MiB,
-			CompactionTotalSize:    DefaultCompactionTotalSize * leveldbopt.MiB,
-			BlockCacheCapacity:     minCache / 2 * leveldbopt.MiB,
-			WriteBuffer:            minCache / 4 * leveldbopt.MiB,
-			Filter:                 filter.NewBloomFilter(DefaultBloomKeyBits),
-			NoSync:                 false,
-		},
+// NewBlockchainStorageBuilder creates the new blockchain storage builder
+func NewBlockchainStorageBuilder(logger hclog.Logger, leveldbBuilder kvdb.LevelDBBuilder) blockchain.StorageBuilder {
```

Comment:
Should it be a high level API instead of the inner module?

Reply:
@DarianShawn  only one place to use, maybe move to up

Reply:
@DarianShawn can't move to the upper level, exist cycle import